src/compiler/parser-decl-struct.bas

```diff
@@ -852,10 +853,19 @@ decl_inner:		'' it's an anonymous inner UDT
 				'' bitfield..
 				hTypeElementDecl( FB_TK_DIM, s, attrib )
 
-			'' it's a field, parse it
-			case else
+			'' it's a field, parse it, 'AS' is a reserved word
+			'' so it's probably OK we don't allow 'TYPE AS' to start
+			'' an inner type declaration and instead expect this is
+			'' going to be 'TYPE AS <datatype>'
+			case FB_TK_AS
 				hTypeElementDecl( FB_TK_DIM, s, attrib )
 
+			'' Otherwise TYPE|UNION starts an inner declaration
+			case else
+				'' allow named UNION|TYPE to be nested in another UNION|TYPE
+				hBeginNesting( s )
+				cTypeDecl( attrib )
```

Comment:
So `cTypeDecl()` can be called recursively now. I noticed one `static` variable in it, `id`, that might need to be changed to normal to support that.

Reply:
Ah, indeed.  `id` is used after the call to `hTypeAdd()` so `id` will need to be on the stack.

Reply:
This should be resolved in [fa021f65ab0b620f7523e4da36e5f443fac6cdae]